controller/src/main/java/com/epam/esm/contoller/GiftCertificateController.java

```diff
@@ -0,0 +1,161 @@
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PatchMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.RestController;
+
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.util.List;
+import java.util.Optional;
+
+@RestController
+@RequestMapping("/certificates")
+public class GiftCertificateController {
+
+    private static final String DEFAULT_LIMIT = "5";
+    private static final String DEFAULT_PAGE = "1";
+    private static final String ERROR_MESSAGE = "error";
+
+    @Autowired
+    private GiftCertificateService service;
+
+    @GetMapping(value = "/")
+    public ResponseEntity<?> retrieveCertificates
+            (@RequestParam(defaultValue = DEFAULT_LIMIT) int limit,
+             @RequestParam(defaultValue = DEFAULT_PAGE) int page,
+             @RequestParam(required = false) String partOfName,
+             @RequestParam(required = false) String partOfDescription,
+             @RequestParam(required = false) String nameOfTag,
+             @RequestParam(required = false) String field,
+             @RequestParam(required = false) String method) {
+        List<GiftCertificateDTO> results;
+        try {
+            if (partOfName != null) {
+                results = service.findByPartOfName(partOfName, limit, page);
+            } else if (partOfDescription != null) {
+                results = service.findByPartOfDescription(partOfDescription, limit, page);
+            } else if (nameOfTag != null) {
+                results = service.findByTag(nameOfTag, limit, page);
+            } else if (field != null && method != null) {
+                results = service.sortByField(field, method, limit, page);
+            } else {
+                results = service.findAll(limit, page);
+            }
+            if (results == null || results.isEmpty()) {
+                ActionHypermediaLinkBuilder builder = new ActionHypermediaLinkBuilder(new ActionHypermedia("not found"));
+                builder.buildRetrieveAllCertificateSelfLink(limit, page, partOfName, partOfDescription, nameOfTag, field, method);
+                return new ResponseEntity<>(builder.getHypermedia(), HttpStatus.NOT_FOUND);
+            }
+            for (int i = 0; i < results.size(); i++) {
+                CertificateLinkBuilder builder = new CertificateLinkBuilder(results.get(i));
+                builder.buildCertificateHypermedia();
+                results.set(i, builder.getHypermedia());
+            }
+            return new ResponseEntity<>(results, HttpStatus.OK);
+        } catch (Throwable e) {
```

Comment:
Relates for all the same pieces of code:
Catching Throwable is not a good practice - https://www.baeldung.com/java-catch-throwable-bad-practice

Reply:
I guess you're trying to perform custom error handling here
Please pay the attention on @ExceptionHandler
https://www.baeldung.com/global-error-handler-in-a-spring-rest-api